Datetime encoder

LeoGrin opened the pull request:
Creates a new encoder which transform datetime columns into several numerical features (year, month, day...). Solves the second part of #233.

LilianBoulard commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,173 @@
+    """
+    This encoder transforms each datetime column into several numeric columns corresponding to temporal features,
+    e.g year, month, day...
+    If one of the feature is constant, this feature is dropped.
```
```suggestion
    Constant extracted features are dropped ; for instance, if the year is always the same in a feature, the extracted "year" column won't be added.
```

LilianBoulard commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,173 @@
+        """
+        Returns clean feature names with format "<column_name>_<new_feature>"
+        if the original data has column names, otherwise with format
+        "<column_indice>_<new_feature>". new_feature is one of ["year", "month",
```
```suggestion
        "<column_index>_<new_feature>". new_feature is one of ["year", "month",
```

LilianBoulard commented on dirty_cat/test/test_datetime_encoder.py:
```diff
@@ -0,0 +1,183 @@
+    assert np.allclose(X_trans, expected_result, equal_nan=True)
+
+    # Datetimes with TZ
+    # If the date are timezone-aware, all the feature extraction should be done
```
```suggestion
    # If the dates are timezone-aware, all the feature extractions should be done
```

LilianBoulard commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,173 @@
+            Data where each column is a datetime feature.
+        Returns
+        -------
+        self
```
```suggestion
        self
            Fitted DatetimeEncoder instance.
```

LilianBoulard commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,173 @@
+        elif feature == "millisecond":
+            return pd.DatetimeIndex(date_series).microsecond.to_numpy()
+        elif feature == "microsecond":
+            return pd.DatetimeIndex(date_series).microsecond.to_numpy()
```
I don't get why it returns the same exact same thing when specifying "millisecond" and "microsecond", is there a reason ?

LeoGrin replied:
Oooops

LilianBoulard commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,173 @@
+        for i in self.features_per_column.keys():
+            prefix = str(i) if self.colnames is None else self.colnames[i]
+            for feature in self.features_per_column[i]:
+                feature_names.append(prefix + "_" + feature)
```
```suggestion
                feature_names.append(f"{prefix}_{feature}")
```

LilianBoulard commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,173 @@
+                              "millisecond": "ms", "microsecond": "us", "nanosecond": "N"}
+
+    def _validate_keywords(self):
+        if self.extract_until not in ["year", "month", "day", "hour", "minute", "second", "millisecond", "microsecond", "nanosecond"]:
+            msg = (
+                """extract_until should be one of ["year", "month", "day", "hour", "minute", "second", "millisecond", 
+                "microsecond", "nanosecond"], got {0}.""".format(
+                    self.extract_until
+                )
+            )
+            raise ValueError(msg)
```
I'd advise using modern python f-strings, which are usually nicer to read :)

```suggestion
        if self.extract_until not in ["year", "month", "day", "hour", "minute", "second", "millisecond", "microsecond", "nanosecond"]:
            raise ValueError(
                f'"extract_until" should be one of ["year", "month", "day", "hour", "minute", '
                f'"second", "millisecond", "microsecond", "nanosecond"], got {self.extract_until}.'
            )
```

LilianBoulard commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,173 @@
+        to_extract = ["year", "month", "day", "hour", "minute", "second", "millisecond", "microsecond", "nanosecond"]
+        self.extract_until = extract_until
+        self._validate_keywords()
+        self.to_extract_full = to_extract[:to_extract.index(extract_until) + 1]
+        self.to_extract_full.append("other")
```
Following the previous comment, you should probably move these lines somewhere in your methods, to avoid raising an IndexError.

LilianBoulard reviewed (changes requested):
Here's bunch of small adjustments, but overall this is very good ! Thank you :)

LeoGrin commented:
Thank you for the review @LilianBoulard !

GaelVaroquaux commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,171 @@
+        self.add_holidays = add_holidays
+        # Some functions need aliases
+        self.word_to_alias = {"year": "Y", "month": "M", "day": "D", "hour": "H", "minute": "min", "second": "S",
+                              "millisecond": "ms", "microsecond": "us", "nanosecond": "N"}
```
The convention in scikit-learn is that the init should do nothing other than set parameters handed in as inputs. One reason is to have a very well understood statefulness of the objects.

These can be set at the beginning of fit.

GaelVaroquaux commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,171 @@
+                              "millisecond": "ms", "microsecond": "us", "nanosecond": "N"}
+
+    def _validate_keywords(self):
+        if self.extract_until not in ["year", "month", "day", "hour", "minute", "second", "millisecond", "microsecond", "nanosecond"]:
```
You could probably define a module-level constant (constants are in ALLCAPS) and reuse it in the file, as is seems that you are using this set quite often.

GaelVaroquaux commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,171 @@
+            Fitted DatetimeEncoder instance.
+        """
+        self._validate_keywords()
+        self.to_extract_full = self.to_extract[:self.to_extract.index(self.extract_until) + 1]
```
If you are never modifying self.to_extract, you can put it as a class variable rather than set it in the init.

LeoGrin replied:
I've made it a module-level constant

GaelVaroquaux commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,171 @@
+        return self
+
+    def transform(self, X, y=None):
+        """ Transform X by replacing each column with corresponding numerical features.
```
I think that you are missing an empty line here for this to be valid docstring formatting.

GaelVaroquaux commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,171 @@
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, ) or (n_samples, 1)
+            The data to transform, where each column is a datetime feature.
```
Also missing an empty line here

GaelVaroquaux commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,171 @@
+        """ Transform X by replacing each column with corresponding numerical features.
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, ) or (n_samples, 1)
```
If the line below is correct, than the line above is incorrect: X can actually be of shape (n_samples, n_features)

GaelVaroquaux commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,171 @@
+        if isinstance(X, pd.DataFrame):
+            self.colnames = X.columns
+        else:
+            self.colnames = None
```
Given that `self.colnames` is determined from the data, it should end with an underscore: `self.colnames_`.

In addition, we prefer to separate words with underscores (to make them easier to read): `self.col_names_`

GaelVaroquaux commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,171 @@
+        else:
+            self.colnames = None
+        X = check_input(X)
+        self.features_per_column = {}  # Features to extract for each column, after removing constant features
```
This should also end with an underscore

GaelVaroquaux commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,171 @@
+            Transformed input.
+        """
+        X = check_input(X)
+        # Create a new dataframe with the extracted features, choosing only features that weren't constant during fit
```
```suggestion
        # Create a new array with the extracted features, choosing only features that weren't constant during fit
```

GaelVaroquaux commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,171 @@
+                feature_names.append(f"{prefix}_{feature}")
+        return feature_names
+
+    def get_feature_names_out(self, input_features=None) -> List[str]:
```
Can you invert get_feature_names and get_feature_names_out: get_feature_names will be removed in a little while, when we no longer depend on old versions of scikit-learn.

GaelVaroquaux commented on dirty_cat/test/test_datetime_encoder.py:
```diff
@@ -0,0 +1,183 @@
+    return np.array([pd.to_datetime(["2020-01-01", "2020-01-02", "2020-01-03"]),
+                  pd.to_datetime(["2021-02-03", "2020-02-04", "2021-02-05"]),
+                  pd.to_datetime(["2022-01-01", "2020-12-25", "2022-01-03"]),
+                  pd.to_datetime(["2023-02-03", "2020-02-04", "2023-02-05"])])
```
In terms of indentation, it would be great if you indented a bit more this list, to align the "pd" to the above "pd" on the first line (comment also applies below in a few places)

GaelVaroquaux commented on dirty_cat/test/test_datetime_encoder.py:
```diff
@@ -0,0 +1,183 @@
+    return np.array([pd.to_datetime(["2020-01-01 10:12:01", "2020-01-02 10:23:00", "2020-01-03 10:00:00"]),
+                  pd.to_datetime(["2021-02-03 12:45:23", "2020-02-04 22:12:00", "2021-02-05 12:00:00"]),
+                  pd.to_datetime(["2022-01-01 23:23:43", "2020-12-25 11:12:00", "2022-01-03 11:00:00"]),
+                  pd.to_datetime(["2023-02-03 11:12:12", "2020-02-04 08:32:00", "2023-02-05 23:00:00"])])
```
PEP8 style guide says that there should be 2 empty lines between first-level functions (comment applies in quite a few places)

GaelVaroquaux reviewed (changes requested):
Nice!

I left a few comments, but none of them are major.

The new object should be added to the list of encoders on the first page (doc/index.rst).

We also need an example demonstrating the new object.

LeoGrin commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,171 @@
+import numpy as np
+import pandas as pd
+from dirty_cat.utils import check_input
+from pandas.tseries.holiday import USFederalHolidayCalendar as calendar
```
Yes I’m not sure. Maybe it’s okay as it is deactivated by default, and the documentation states that it only supports US holidays for now? Otherwise, we can maybe use https://github.com/dr-prodigy/python-holidays. Do you think a new dependency is worth it?

LeoGrin commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,171 @@
+    "millisecond", "microsecond", "nanosecond"}, default="hour"
+        Extract up to this granularity, and gather the rest into the "other" feature.
+        For instance, if you specify "day", only "year", "month", "day" and "other" features will be created.
+        The "other" feature will be a numerical value expressed in the "extract_until" unit.
```
Hmm interesting. I get you point, but here's why I chose to do it this way:
- I fear that full_time_to_epoch would be very collinear with the highest-level feature, hurting interpretability. For a worst case scenario, we can imagine a dataset of different hours during the same day for different years.
- It’s true that with my choice, the learner has to learn the weird algebra. But with your proposition, the learner would still need to learn this weird algebra if it wants to use my `other` feature (which may be more important for prediction since it’s information which isn’t contained in the other variables), because `other` = full_time_to_epoch - (year - 1970) * 365 * 24 * 3600 + …
Maybe using both could indeed make the learning easier, but wouldn't you worry about collinearity?
All in all I'm not sure.

GaelVaroquaux replied:
I do machine learning. I don't worry about collinearity :).

It does not hurt for prediction. I can hurt for interpretability.

GaelVaroquaux replied:
My reason for favoring the full_time_to_epoch is that I think that, in general, it is more likely to be useful than the other end, and we are more likely to have to reconstruct it.

GaelVaroquaux commented:
> Do you think a new dependency is worth it?

No, I don't think that a new dependency is worth it.

I'm wondering if we should simply drop this feature (less features = less code = less problems :D )

LeoGrin commented on dirty_cat/datetime_encoder.py:
```diff
@@ -0,0 +1,171 @@
+    Parameters
+    ----------
+    extract_until : {"year", "month", "day", "hour", "minute", "second",
+    "millisecond", "microsecond", "nanosecond"}, default="hour"
```
It was a mess indeed :)

LeoGrin commented:
> I'm wondering if we should simply drop this feature
I've removed it.

GaelVaroquaux reviewed (changes requested):
I think that:
- I still prefer to have the full time to epoch
- We are lacking an example (examples are crucial). We probably need to add a new example about this specific topic.

LeoGrin commented:
Additions:
- Replaced the "other" feature by the "full" features, which contains the full time to epoch (see above discussion with @GaelVaroquaux)
- Made the DatetimeEncoder the default for datetime columns in the SuperVectorizer
- Added a simple example
- Small fixes

LeoGrin commented:
Apparently using fetch_traffic_violations() in the example takes too much memory for circleCI

LeoGrin commented:
I've changed the example to use another dataset, I think the PR is ready for review @LilianBoulard @GaelVaroquaux @jovan-stojanovic

GaelVaroquaux commented on examples/06_datetime_encoder.py:
```diff
@@ -0,0 +1,103 @@
+
+sup_vec = SuperVectorizer()
+X_ = sup_vec.fit_transform(X)
+feature_names = sup_vec.get_feature_names_out()
+print(feature_names)
```
```suggestion
sup_vec.get_feature_names_out()
```
I think that the code will be easier to read and the formatting of the example will be nicer like this.

GaelVaroquaux commented on examples/06_datetime_encoder.py:
```diff
@@ -0,0 +1,103 @@
+# If we want the day of the week, we can just replace SuperVectorizer's default
+sup_vec = SuperVectorizer(datetime_transformer=DatetimeEncoder(add_day_of_the_week=True))
+X_ = sup_vec.fit_transform(X)
+feature_names = sup_vec.get_feature_names_out()
+print(feature_names)
```
```suggestion
sup_vec.get_feature_names_out()
```

GaelVaroquaux commented on examples/06_datetime_encoder.py:
```diff
@@ -0,0 +1,103 @@
+breaks down each datetime features into several numerical features, by extracting relevant information from the
+datetime features, such as the month, the day of the week, the hour of the day, etc. Used in
+the SuperVectorizer, which automatically detects the datetime features, the DatetimeEncoder allows
+to handle datetime features easily.
```
Would it be possible to do a small prediction model at the end, add a section, to showcase this in a prediction pipeline?

For didactic purposes it would be important to have not only the cross-validation (using https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html ) but also the plot of the extrapolated time series (compared to the real one)

GaelVaroquaux commented on dirty_cat/datetime_encoder.py:
```diff
@@ -20,9 +20,9 @@ class DatetimeEncoder(TransformerMixin, BaseEstimator):
-        Extract up to this granularity, and gather the rest into the "other" feature.
-        For instance, if you specify "day", only "year", "month", "day" and "other" features will be created.
-        The "other" feature will be a numerical value expressed in the "extract_until" unit.
+        Extract up to this granularity. If all features have not been extracted, add the "full" feature, which contains
```
Maybe call it "total_time" rather than "full". I had not understood what "full" meant.

GaelVaroquaux reviewed (changes requested):
A few changes requested. Most are minor. I hope that the prediction example won't ask for too much work.

LeoGrin commented:
Thank you for the comments! I've added a new section to the example

GaelVaroquaux commented:
Oups, I just saw that there is a "examples/.DS_Store" that needs to be removed. You can add it in the gitignore if it's helpful.

GaelVaroquaux commented on examples/06_datetime_encoder.py:
```diff
@@ -0,0 +1,159 @@
+
+###############################################################################
+# Transforming the input data
+# ----------------------------
```
```suggestion
# ..........................
```
Decreasing the number of top-level sections (to make the table of content in the side bar more readable)

GaelVaroquaux commented on examples/06_datetime_encoder.py:
```diff
@@ -0,0 +1,159 @@
+feature_names
+
+###############################################################################
+# Features importance
```
I think that I would move the feature importance section below the prediction example. It feels more natural to understand feature importances in such contexte

GaelVaroquaux reviewed (commented):
Still cosmetic comments on the example :)

GaelVaroquaux commented:
Hum, I merged this, but I hadn't seen that the lines are too long (I'll fix it, I wanted to do cosmetics on the example anyhow).

It does have a practical consequence: it forces horizontal scrolling in the example, which is bad for readability.

GaelVaroquaux commented:
> This actually made me wonder: in which case should I use a class attribute vs a module-level constant?

Use a class attribute if you might want to override it in a subclass.
